File: solutions/ocaml/minesweeper/1/minesweeper.ml
let cell_increase board row col =
  if row < 0 || col < 0 || row >= Array.length board || col >= Bytes.length board.(row) then
    ()
  else
    match Bytes.get board.(row) col with
    | '*' -> ()
    | ' ' -> Bytes.set board.(row) col '1'
    | c -> Bytes.set board.(row) col (Char.chr (Char.code c + 1))

let annotate field =
  let board = Array.map Bytes.of_string (Array.of_list field) in
  for row = 0 to Array.length board - 1 do
    for col = 0 to Bytes.length board.(row) - 1 do
      if Bytes.get board.(row) col = '*' then (
        cell_increase board (row - 1) (col - 1);
        cell_increase board (row - 1) col;
        cell_increase board (row - 1) (col + 1);
        cell_increase board row (col - 1);
        cell_increase board row (col + 1);
        cell_increase board (row + 1) (col - 1);
        cell_increase board (row + 1) col;
        cell_increase board (row + 1) (col + 1)
      )
    done
  done;
  Array.map Bytes.to_string board |> Array.to_list

File: solutions/typescript/minesweeper/1/minesweeper.ts

function cellIncrease(board: string[][], row: number, col: number) {
    if (row < 0 || col < 0 || row >= board.length || col >= board[row].length) {
        return;
    }
    switch (board[row][col]) {
        case '*':
            break;
        case ' ':
            board[row][col] = '1';
            break;
        default:
            board[row][col] = String(Number(board[row][col]) + 1);
            break;
    }
}

export function annotate(field: string[]): string[] {
    const board: string[][] = [];
    for (const row of field) {
        board.push([...row]);
    }
    for (let row = 0; row < field.length; row++) {
        for (let col = 0; col < field[row].length; col++) {
            if (field[row][col] === '*') {
                cellIncrease(board, row - 1, col - 1);
                cellIncrease(board, row - 1, col);
                cellIncrease(board, row - 1, col + 1);
                cellIncrease(board, row, col - 1);
                cellIncrease(board, row, col + 1);
                cellIncrease(board, row + 1, col - 1);
                cellIncrease(board, row + 1, col);
                cellIncrease(board, row + 1, col + 1);
            }
        }
    }
    const result: string[] = [];
    for (const row in board) {
        result.push(board[row].join(''));
    }
    return result;
}
